lib/THC/THCTensorMath.cuh

```diff
@@ -23,4 +23,31 @@ __global__ void THCTensor_copyToDiagonal(T* a, T* b, ptrdiff_t start, ptrdiff_t
   }
 }
 
+#define CAT_ARRAY_KERNEL_BATCH_SIZE 2
```

Comment:
re: constant memory, you could be calling this concurrently from different streams, so you need a different region of constant memory for each stream.

instead of global memory, i think i'd just specialize the kernel further to only handle contiguous inputs of 4 dims or less, so you don't need the stride array at all, and you don't need MAX_DIMS sizes.